src/render/src/macros/mod.rs

```diff
@@ -31,7 +31,107 @@ macro_rules! gfx_format {
 }
 
 
-/// Defines vertex, constant and pipeline formats in one block
+/// Defines vertex, constant and pipeline formats in one block.
+///
+/// # Example
+///
+/// ```{.rust}
+/// #[macro_use]
+/// extern crate gfx;
+///
+/// gfx_defines! {
+///     vertex Vertex {
+///         pos: [f32; 4] = "a_Pos",
+///         tex_coord: [f32; 2] = "a_TexCoord",
+///     }
+///     
+///     constant Locals {
+///         transform: [[f32; 4]; 4] = "u_Transform",
+///     }
+///
+///     pipeline pipe {
+///         vbuf: gfx::VertexBuffer<Vertex> = (),
+///         // Global buffers are added for compatibility when
+///         // constant buffers are not supported.
+///         transform: gfx::Global<[[f32; 4]; 4]> = "u_Transform",
+///         locals: gfx::ConstantBuffer<Locals> = "Locals",
+///         color: gfx::TextureSampler<[f32; 4]> = "t_Color",
+///         out_color: gfx::RenderTarget<gfx::format::Rgba8> = "Target0",
+///         out_depth: gfx::DepthTarget<gfx::format::DepthStencil> = 
+///             gfx::preset::depth::LESS_EQUAL_WRITE,
+///     }
+/// }
+///
+/// impl Vertex {
+///     fn new(p: [i8; 3], tex: [i8; 2]) -> Vertex {
+///         Vertex {
+///             pos: [p[0] as f32, p[1] as f32, p[2] as f32, 1.0f32],
+///             tex_coord: [tex[0] as f32, tex[1] as f32],
+///         }
+///     }
+/// }
+///
+/// fn main() {
+///     let vertex_data = [
+///         Vertex::new([-1, -1, 1], [0, 0]),
+///         Vertex::new([ 1, -1, 1], [1, 0]),
+///         // Add more vertices..
+///     ];
+/// }
+/// ```
+/// `vertex` and `constant` structures defined with `gfx_defines!`
+/// can be extended with attributes:
+///
+/// ```{.rust}
+/// #[macro_use]
+/// extern crate gfx;
+///
+/// gfx_defines! {
+///     #[derive(Default)]
+///     vertex Vertex {
+///         pos: [f32; 4] = "a_Pos",
+///         tex_coord: [f32; 2] = "a_TexCoord",
+///     }
+/// }
+///
+/// fn main() {
+///     let vertex = Vertex::default();
+///     assert_eq!(vertex.pos[0], 0f32);
+///     assert_eq!(vertex.tex_coord[0], 0f32);
+/// }
+/// ```
+///
+/// # `pipe`
+///
+/// The `pipeline state object` or `pso` can consist of the following
+/// `pso` components:
+///
+/// - A single [vertex buffer](pso/buffer/type.VertexBuffer.html) component to hold the vertices.
+/// - A single [instance buffer](pso/buffer/type.InstanceBuffer.html) component.
+/// - Single or multiple [constant buffer](pso/buffer/struct.ConstantBuffer.html) components. (DX11 and OpenGL3)
+/// - Single or multiple [global buffer](pos/buffer/struct.Global.html) components.
+/// - Single or multiple [samplers](pos/resource/struct.Sampler.html) 
+///   such as [texture samplers](pso/resource/struct.TextureSampler.html].
```

Comment:
A `TextureSampler` component is not an example of a sampler, it's a helper that is basically "ShaderResource" + "Sampler", which is useful for OpenGL where each texture comes with a sampler by default.